### PR TITLE
Feature/PromiseKit - Renaming proposal + `Any` and `Race`

### PR DESCRIPTION
**Pull request by tonikocjan**
This ticket proposes the following naming changes:
  1. `combine` => `all`
  2. `onSuccess` => `then`
  3. `onFailure` => `catch`
  4. `observe` => `finally`

I think the proposed names better convey the underlying meaning and they more closely follow the _Promise standard_. The old methods were deprecated only, which means that existing code using PromiseKit should still compile.

Additionally, 
  - `any` (a dual of `all`), which allows the user to combine several promises allowing some to fail, and 
  - `race`, which returns a promise that fulfills or rejects as soon as one of the promises in the collection fulfills or rejects, with the result from that promise

are implemented (together with [tests](https://github.com/poviolabs/PovioKit/pull/119/files#diff-45eb96e7c84e64adf2ac03ce2fec2a6ad728250229a154f617b0c5358d7e3c57)).

### CODE REVIEW

**Review by ArminSpahic — APPROVED** [review]

File: Sources/PromiseKit/Any.swift [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  guard !promises.isEmpty else {
+    return .error(NSError(domain: "com.poviokit.promisekit", code: 101, userInfo: nil))
+  }
+  
+  return .init { seal in
+    let barrier = DispatchQueue(label: "combineQueue", attributes: .concurrent)
+    for promise in promises {
+      promise.finally { result in
+        barrier.async(flags: .barrier) {
+          guard promises.allSatisfy({ $0.isResolved }) else { return }
```
Comment (borut-t) [inline Sources/PromiseKit/Any.swift:31]: `isResolved` flags what?

File: Sources/PromiseKit/Race.swift [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+public func race<T, C: Collection>(
+  on dispatchQueue: DispatchQueue = .main,
+  promises: C
+) -> Promise<T> where C.Element == Promise<T> {
+  guard !promises.isEmpty else {
+    return .error(NSError(domain: "com.poviokit.promisekit", code: 101, userInfo: nil))
+  }
+  
+  return .init { seal in
+    let barrier = DispatchQueue(label: "combineQueue", attributes: .concurrent)
```
Comment (borut-t) [inline Sources/PromiseKit/Race.swift:25]: Why do you name queue "barrier" 🤔

**Review by borut-t — APPROVED** [review]
Great work and I think renaming following the Promise naming convention makes sense. Can you please update documentation and add examples for added options?

File: Sources/PromiseKit/Any.swift (hunk 1 above)
Comment (tonikocjan) [reply Sources/PromiseKit/Any.swift:31]: `isResolved` means is that the promise was fulfilled with a value.

File: Sources/PromiseKit/Race.swift (hunk 2 above)
Comment (tonikocjan) [reply Sources/PromiseKit/Race.swift:25]: https://developer.apple.com/documentation/dispatch/dispatch_barrier

**Comment by tonikocjan** [comment]
>  Can you please update documentation and add examples for added options?

Absolutely.

**Pull request merged**